lib/streamlit/elements/lib/subtitle_utils.py

```diff
@@ -0,0 +1,177 @@
+    except UnicodeDecodeError:
+        # If it's not valid utf-8, it's probably not a valid SRT file
+        return False
+
+    # Split the header into lines and process them
+    lines = header_str.split("\n")
+
+    # Check for the pattern of an SRT file: digit(s), newline, timestamp
+    if len(lines) >= 2 and lines[0].isdigit():
+        match = re.search(SRT_VALIDATION_REGEX, lines[1])
+        if match:
+            return True
+
+    return False
+
+
+def _srt_to_vtt(srt_data: str | bytes) -> bytes:
+    """
+    Convert subtitles from SubRip (.srt) format to WebVTT (.vtt) format.
+    This function accepts the content of the .srt file either as a string
+    or as a BytesIO stream.
+    Parameters
+    ----------
+    srt_data : str or bytes
+        The content of the .srt file as a string or a bytes stream.
+    Returns
+    -------
+    bytes
+        The content converted into .vtt format.
+    """
+
+    # If the input is a bytes stream, convert it to a string
+    if isinstance(srt_data, bytes):
+        # Decode the bytes to a UTF-8 string
+        try:
+            srt_data = srt_data.decode("utf-8")
+        except UnicodeDecodeError as e:
+            raise ValueError("Could not decode the input stream as UTF-8.") from e
+    if not isinstance(srt_data, str):
+        # If it's not a string by this point, something is wrong.
+        raise TypeError(
+            f"Input must be a string or a bytes stream, not {type(srt_data)}."
+        )
+
+    # Replace SubRip timing with WebVTT timing
+    vtt_data = re.sub(SRT_CONVERSION_REGEX, r"\1.\2", srt_data)
+
+    # Add WebVTT file header
+    vtt_content = "WEBVTT\n\n" + vtt_data
+    # Convert the vtt content to bytes
+    vtt_content = vtt_content.strip().encode("utf-8")
+
+    return vtt_content
+
+
+def _handle_string_or_path_data(data_or_path: str | Path) -> bytes:
+    """Handles string data, either as a file path or raw content."""
+    if os.path.isfile(data_or_path):
+        path = Path(data_or_path)
+        file_extension = path.suffix.lower()
+
+        if file_extension not in SUBTITLE_ALLOWED_FORMATS:
+            raise ValueError(
+                f"Incorrect subtitle format {file_extension}. Subtitles must be in "
+                f"one of the following formats: {', '.join(SUBTITLE_ALLOWED_FORMATS)}"
+            )
+        with open(data_or_path, "rb") as file:
+            content = file.read()
+        return _srt_to_vtt(content) if file_extension == ".srt" else content
+    elif isinstance(data_or_path, Path):
+        raise ValueError(f"File {data_or_path} does not exist.")
+
+    content_string = data_or_path.strip()
+
+    if content_string.startswith("WEBVTT") or content_string == "":
+        return content_string.encode("utf-8")
+    elif _is_srt(content_string):
+        return _srt_to_vtt(content_string)
+    raise ValueError("The provided string neither matches valid VTT nor SRT format.")
```

Comment:
Do you know if these `ValueErrors` ever reached in practice? In my experiments, trying to enter misformatted strings has been leading me to just `StreamlitAPIException: Failed to process the provided subtitle:`